### PR TITLE
implement radiusd::radius_xlat in rlm_perl

### PR DESCRIPTION
**Pull request by lytboris**
Main goal is to get access to things like `%{client:group}` inside perl scripts

### CODE REVIEW

**Comment by arr2036** [comment]
Get rid of the radius and just do radiusd::xlat

**Comment by arr2036** [comment]
and use radius_axlat instead of a static buffer

File: doc/ChangeLog [hunk 1]
```diff
@@ -22,6 +22,8 @@ FreeRADIUS 3.1.0 Thu 26 Nov 2015 15:48:14 EDT urgency=medium
 	* TLV Attributes can now be nested to nearly any depth.
 	* radsnmp client which can listen on Net-SNMP "pass persist"
 	  and translate that into queries to the server.
+	* rlm_perl: radiusd::radius_xlat to evaluate xlat string
```
Comment (arr2036) [inline doc/ChangeLog]: Just use radiusd::xlat, no need for the extra qualification

File: src/modules/rlm_perl/rlm_perl.c [hunk 2]
```diff
@@ -296,6 +296,40 @@ static XS(XS_radiusd_radlog)
...
+	dXSARGS;
+	char *in_str;
+	char expanded[1024]; /* XXX FIXME Dynamic allocation in case slen >= sizeof(expanded) ? */
+	ssize_t slen;
+	SV *rad_requestp_sv;
+	REQUEST *request;
+
+	if (items != 1) croak("Usage: radiusd::radius_xlat(string)");
+
+	rad_requestp_sv = get_sv("RAD___REQUESTP", 0);
```
Comment (arr2036) [inline src/modules/rlm_perl/rlm_perl.c]: Do it like this instead:
- https://github.com/FreeRADIUS/freeradius-server/blob/v4.0.x/src/modules/rlm_lua/lua.c#L40
- https://github.com/FreeRADIUS/freeradius-server/blob/v4.0.x/src/modules/rlm_lua/lua.c#L247
- https://github.com/FreeRADIUS/freeradius-server/blob/v4.0.x/src/modules/rlm_lua/lua.c#L770

Many many many fewer cycles.

File: src/modules/rlm_perl/rlm_perl.c [hunk 3]
```diff
@@ -296,6 +296,40 @@ static XS(XS_radiusd_radlog)
...
+	if (rad_requestp_sv == NULL) croak("Can not evalue xlat, RAD___REQUESTP is not set!");
+
+	request = INT2PTR(REQUEST *, SvIV(rad_requestp_sv));
+
+	in_str = (char *) SvPV(ST(0), PL_na);
+	memset(expanded, 0, sizeof(expanded));
+	slen = radius_xlat(expanded, sizeof(expanded), request, in_str, NULL, NULL);
```
Comment (arr2036) [inline src/modules/rlm_perl/rlm_perl.c]: use ``radius_axlat`` so the strings aren't truncated.

File: src/modules/rlm_perl/rlm_perl.c [hunk 4]
```diff
@@ -296,6 +296,40 @@ static XS(XS_radiusd_radlog)
...
+ *	Now users are able to get data that is accessible only via xlat
+ *	e.g. %{client:...}
+ *	Call syntax is radiusd::radius_xlat(string), string will be handled the
+ *	same way it is described in EXPANSIONS section of man unlang
+ */
+static XS(XS_radiusd_radius_xlat)
+{
+	dXSARGS;
+	char *in_str;
+	char expanded[1024]; /* XXX FIXME Dynamic allocation in case slen >= sizeof(expanded) ? */
```
Comment (arr2036) [inline src/modules/rlm_perl/rlm_perl.c]: Yes..

**Comment by lytboris** [comment]
Pushed updated version. It is not tested in prod though, will do it on week following (on 3.0.x with __THREAD REQUEST *rlm_perl_request).

**Comment by arr2036** [comment]
Looks good, thanks!

**Comment by arr2036** [comment]
v3.0.x does have thread_local macros which fallback to mutexes. Honestly any compiler written in the last five years will have support for thread local variables, even if it's not using the C11 keyword.

**Pull request merged**